internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,349 @@
+package sshsession
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/juju/errors"
+	jujussh "github.com/juju/utils/v3/ssh"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/rpc/params"
+)
+
+var (
+	// ControllerSSHUser is the user that will connect to the controller's embedded
+	// SSH server.
+	ControllerSSHUser = "reverse-ssh"
+	// We use authorized_keys2 as it's a default configured alternative file to
+	// authorized_keys. Our existing SSH implementation was hardcoded against authorized_keys
+	// and using this file allows us to not interfere with existing setups. The user may
+	// still add another file to the "AuthorizedKeysFile" directive in sshd_config.
+	//
+	// TODO(ale8k): At somepoint, update juju documentation to explain should they wish
+	// to ssh directly to a machine, this is the way to do so.
+	AuthorizedKeysFile = "authorized_keys2"
+)
+
+// FacadeClient holds the facade methods for the SSH session worker.
+type FacadeClient interface {
+	// WatchSSHConnRequest creates a watcher and returns its ID for watching changes.
+	WatchSSHConnRequest(machineId string) (watcher.StringsWatcher, error)
+
+	// GetSSHConnRequest returns a ssh connection request by its connection request ID.
+	GetSSHConnRequest(arg string) (params.SSHConnRequest, error)
+}
+
+// WorkerConfig encapsulates the configuration options for
+// instantiating a new lease ssh session worker.
+type WorkerConfig struct {
+	Logger           Logger
+	MachineId        string
+	FacadeClient     FacadeClient
+	ConnectionGetter ConnectionGetter
+}
+
+// Validate checks whether the worker configuration settings are valid.
+func (cfg WorkerConfig) Validate() error {
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.MachineId == "" {
+		return errors.NotValidf("empty MachineId")
+	}
+	if cfg.FacadeClient == nil {
+		return errors.NotValidf("nil FacadeClient")
+	}
+	if cfg.ConnectionGetter == nil {
+		return errors.NotValidf("nil ConnectionGetter")
+	}
+
+	return nil
+}
+
+// sshSessionWorker is a worker that enables SSH connections to a machine.
+// It watches for incoming SSH connections to the controller's SSH server over
+// a facade, and when notified initiates a connection to the controller's SSH server.
+// When the controller's SSH server receives this connection, it initiates its own
+// SSH connection down the tunnel and the sshSessionWorker pipes it to the SSHD
+// on the machine.
```

Comment:
I feel like I just read this, in the doc.go file and now again here? Should we keep one copy?